dogsdogsdogs/src/operators/half_join.rs

```diff
@@ -154,81 +161,105 @@ where
     // Stash for (time, diff) accumulation.
     let mut output_buffer = Vec::new();
 
-    stream.inner.binary_frontier(&arrangement_stream, exchange, Pipeline, "HalfJoin", move |_,_| move |input1, input2, output| {
+    stream.inner.binary_frontier(&arrangement_stream, exchange, Pipeline, "HalfJoin", move |_,info| {
 
-        // drain the first input, stashing requests.
-        input1.for_each(|capability, data| {
-            data.swap(&mut buffer);
-            stash.entry(capability.retain())
-                 .or_insert(Vec::new())
-                 .extend(buffer.drain(..))
-        });
+        // Acquire an activator to reschedule the operator when it has unfinished work.
+        use timely::scheduling::Activator;
+        let activations = stream.scope().activations();
+        let activator = Activator::new(&info.address[..], activations);
 
-        // Drain input batches; although we do not observe them, we want access to the input
-        // to observe the frontier and to drive scheduling.
-        input2.for_each(|_, _| { });
+        move |input1, input2, output| {
 
-        if let Some(ref mut trace) = arrangement_trace {
+            // drain the first input, stashing requests.
+            input1.for_each(|capability, data| {
+                data.swap(&mut buffer);
+                stash.entry(capability.retain())
+                    .or_insert(Vec::new())
+                    .extend(buffer.drain(..))
+            });
 
-            for (capability, proposals) in stash.iter_mut() {
+            // Drain input batches; although we do not observe them, we want access to the input
+            // to observe the frontier and to drive scheduling.
+            input2.for_each(|_, _| { });
 
-                // defer requests at incomplete times.
-                // TODO: Verify this is correct for TOTAL ORDER.
-                if !input2.frontier.less_equal(capability.time()) {
+            // Local variables to track if and when we should exit early.
+            // The rough logic is that we fully process inputs and set their differences to zero,
+            // stopping at any point. We clean up all of the zeros in buffers that did any work,
+            // and reactivate at the end if the yield function still says so.
+            let mut yielded = false;
+            let timer = std::time::Instant::now();
+            let mut work = 0;
 
-                    let mut session = output.session(capability);
+            if let Some(ref mut trace) = arrangement_trace {
 
-                    // Sort requests by key for in-order cursor traversal.
-                    consolidate_updates(proposals);
+                for (capability, proposals) in stash.iter_mut() {
 
-                    let (mut cursor, storage) = trace.cursor();
+                    // Avoid computation if we should already yield.
+                    // TODO: Verify this is correct for TOTAL ORDER.
```

Comment:
@frankmcsherry Did you verify this?

Reply:
Not yet. This is an old comment though. I think it is green because the indentation level changed.